miden-crypto/src/dsa/ecdsa_k256/mod.rs

```
@@ -0,0 +1,311 @@
+    /// generated by the other party.
+    pub fn get_shared_secret(&self, pk_e: EphemeralPublicKey) -> SharedSecret {
+        let shared_secret_inner = diffie_hellman(self.inner.as_nonzero_scalar(), pk_e.as_affine());
+
+        SharedSecret::new(shared_secret_inner)
+    }
+}
+
+// PUBLIC KEY
+// ================================================================================================
+
+/// Public key for ECDSA signature verification over secp256k1 curve.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PublicKey {
+    pub(crate) inner: VerifyingKey,
+}
+
+impl PublicKey {
+    /// Returns a commitment to the public key using the RPO256 hash function.
+    pub fn to_commitment(&self) -> Word {
+        <Self as SequentialCommit>::to_commitment(self)
+    }
+
+    /// Verifies a signature against this public key and message.
+    pub fn verify(&self, message: Word, signature: &Signature) -> bool {
+        let message_digest = hash_message(message);
+        let signature_inner = k256::ecdsa::Signature::from_scalars(*signature.r(), *signature.s());
+
+        match signature_inner {
+            Ok(signature) => self.inner.verify_prehash(&message_digest, &signature).is_ok(),
+            Err(_) => false,
+        }
+    }
+
+    /// Recovers from the signature the public key associated to the secret key used to sign the
+    /// message.
+    pub fn recover_from(message: Word, signature: &Signature) -> Result<Self, PublicKeyError> {
+        let message_digest = hash_message(message);
+        let signature_data = k256::ecdsa::Signature::from_scalars(*signature.r(), *signature.s())
+            .expect("could not build the signature from the scalars");
+
+        let verifying_key = k256::ecdsa::VerifyingKey::recover_from_prehash(
+            &message_digest,
+            &signature_data,
+            RecoveryId::from_byte(signature.v()).expect("invalid recovery id"),
+        )
+        .map_err(|_| PublicKeyError::RecoveryFailed)?;
```

Could we add `SAFETY` comments explaining why building signature from scalars and building recovery ID from bytes should never fail?

---

Actually, it could fail if the signature is malformed. Hence we probably need to return an error in this case